homeassistant/components/sql/config_flow.py

```diff
@@ -0,0 +1,143 @@
+"""Adds config flow for SQL integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import sqlalchemy
+from sqlalchemy.orm import scoped_session, sessionmaker
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.recorder import CONF_DB_URL, DEFAULT_DB_FILE, DEFAULT_URL
+from homeassistant.const import CONF_NAME, CONF_UNIT_OF_MEASUREMENT, CONF_VALUE_TEMPLATE
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import TemplateError
+import homeassistant.helpers.config_validation as cv
+import homeassistant.helpers.template as template_helper
+
+from .const import CONF_COLUMN_NAME, CONF_QUERY, DB_URL_RE, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def redact_credentials(data: str) -> str:
+    """Redact credentials from string data."""
+    return DB_URL_RE.sub("//****:****@", data)
+
+
+def validate_sql_select(value: str) -> str | None:
+    """Validate that value is a SQL SELECT query."""
+    if not value.lstrip().lower().startswith("select"):
+        return None
+    return value
+
+
+def validate_template(value: str | None) -> template_helper.Template | None:
+    """Validate template syntax."""
+    template_value = template_helper.Template(str(value))
+    try:
+        template_value.ensure_valid()
+    except TemplateError:
+        return None
+    return template_value
+
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_DB_URL): cv.string,
+        vol.Required(CONF_COLUMN_NAME): cv.string,
+        vol.Required(CONF_QUERY): cv.string,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
+        vol.Optional(CONF_VALUE_TEMPLATE): cv.string,
+    }
+)
+
+
+class SQLConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for SQL integration."""
+
+    VERSION = 1
+
+    entry: config_entries.ConfigEntry
+    hass: HomeAssistant
+
+    async def validate_db(self, db_in: str | None) -> str | None:
+        """Validate database from user input."""
+
+        if not (db_url := db_in):
+            db_url = DEFAULT_URL.format(
+                hass_config_path=self.hass.config.path(DEFAULT_DB_FILE)
+            )
+
+        sess = None
+        try:
+            engine = sqlalchemy.create_engine(db_url)
+            sessmaker = scoped_session(sessionmaker(bind=engine))
+
+            # Run a dummy query just to test the db_url
+            sess = sessmaker()
+            sess.execute("SELECT 1;")
+
+        except sqlalchemy.exc.SQLAlchemyError as err:
+            _LOGGER.error(
+                "Couldn't connect using %s DB_URL: %s",
+                redact_credentials(db_url),
+                redact_credentials(str(err)),
+            )
+            return None
+        finally:
+            if sess:
+                sess.close()
```

Comment:
This needs to happen in the executor

Reply:
You mean to move it up to `try`?

Reply:
This is doing I/O in the event loop.  

Move the try/except/finally block into a new function and run it in the executor with `await self.hass.async_add_executor_job`